### PR TITLE
[WIP] [1140] use longest option argument again

### PR DESCRIPTION
**Pull request by altendky**
#1140

### CODE REVIEW

**Comment by altendky** [comment]
WIP for:
* [ ] Doc fixes

**Review by amiryal — APPROVED** [review]
My only feedback, other than LGTM 👍, is regarding commit messages. Maybe you could provide more context, i.e. what the change actually does and why, and possibly squash some or all of them together.

**Comment by amiryal** [comment]
Just wanted to echo your earlier https://github.com/pallets/click/issues/1140#issuecomment-432291544 here as well:
> It's kind of tempting to deprecate this 'feature' and just say 'if you have multiple of the longest type of parameter, specify the name'. Or maybe provide a marker so you don't have to repeat. Maybe `'+--the-default'` or somesuch? I dunno, explicit over implicit and all. This just seems like a bit of a fiddly mess with little benefit.
> Sorry, make it backwards compatible but fix the situation by deprecating the implicit order-based selection.

:+1: 
I fully agree, Zen of Python and all.

File: tests/test_options.py [hunk 1]
```diff
@@ -462,14 +462,15 @@ def cli_alt(warnings):
     result = runner.invoke(cli_alt, ['-w'])
     assert result.output == 'True\n'
 
+
 @pytest.mark.parametrize('option_args,expected', [
     (['--aggressive', '--all', '-a'], 'aggressive'),
-    (['--first', '--second', '--third', '-a', '-b', '-c'], 'first'),
-    (['--apple', '--banana', '--cantaloupe', '-a', '-b', '-c'], 'apple'),
+    (['--first', '--second', '--third', '-a', '-b', '-c'], 'second'),
```
Comment (sirosen) [inline tests/test_options.py]: This makes me so dizzy to read. 😂 

I understand that we're going back to documented behavior, but this is pretty much the perfect example of why I think [this comment of yours is spot on](https://github.com/pallets/click/issues/1140#issuecomment-432291544). We should really just tell people to write down what they mean...

**Review by sirosen — COMMENTED** [review]
@altendky, is this still a work in progress?

Since this is the course of action which you and @davidism agreed on, and the changes appear to be complete to me, I'd like to merge.

I agree that your commit messages are a bit minimal -- personally, I'd prefer them either squashed or fleshed out with explanation -- but all of the important discussion is in GitHub anyway.

**Comment by altendky** [comment]
After https://github.com/pallets/click/issues/1140#issuecomment-432304387 and https://github.com/pallets/click/issues/1140#issuecomment-432307145 I figured we should figure out where we are going since we have no totally obvious thing to revert to.  If we want to 'just revert' for now I would think we would want the previous functionality which would be 'last long' not 'longest' like this implements.

**Comment by sirosen** [comment]
Ah, I thought after [this comment that we should revert the behavior](https://github.com/pallets/click/issues/1140#issuecomment-429924902) that this reverts #794.

Reading that PR again, it looks like we switched from "last long opt" to "first long opt" but never updated the docs? I didn't realize we never had the also-documented "longest option" behavior.

Sorry for jumping in on this without properly grasping the situation.

Now that I do, I actually think this PR is probably a bad call -- we should either keep the new behavior in 7.0 and document it, or revert to pre #794 behavior and note it as a bugfix in the 7.1 changelog.

IMO, keeping the new behavior and just updating the docs is the cheapest/easiest solution. If someone showed up as a new user of 7.0, then any behavior change appears to be breaking.

**Comment by altendky** [comment]
I only left this open to avoid thrash just in case... But yeah, this is likely just totally dead.

**Pull request closed without merging**